File: vvd-common/src/main/java/mikufan/cx/vvd/common/label/FailedObject.java
package mikufan.cx.vvd.common.label;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.ToString;
import lombok.experimental.FieldDefaults;
import lombok.experimental.SuperBuilder;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;

/**
 * To support non-blocking debugging purposes. <br/>
 * App can use this class to collect or write all failure into one place, <br/>
 * so that when long running app is finished, we can get a clear look of all failure we have <br/>
 * @author CX无敌
 * @date 2020-12-19
 */
@Getter @ToString
@SuperBuilder(toBuilder = true)
@FieldDefaults(makeFinal = true, level = AccessLevel.PROTECTED)
public abstract class FailedObject<T> {

  /**
   * rename to object when the new list provider is done,
   * right now this name is for competitive reason
   */
  @NotNull T failedObj;
  @NotBlank String reason;

}

File: vvd-common-kt/src/main/kotlin/mikufan/cx/vvd/commonkt/label/FailedObject.kt
package mikufan.cx.vvd.commonkt.label

/**
 * To support non-blocking debugging purposes.
 *
 * App can use this class to collect or write all failure into one place,
 *
 * so that when long running app is finished, we can get a clear look of all failure we have.
 *
 * @date 2021-06-12
 * @author CX无敌
 */
data class FailedObject<T>(
  val obj: T,
  val exception: Exception? = null,
  val reason: String? = null
)
